File: 5. Validations and transformations for backend engineers/code/typescript/example_5.ts
import { z } from 'zod';

export const SignupSchema = z.object({
  password: z.string().min(8),
  passwordConfirmation: z.string(),
  married: z.boolean(),
  partner: z.string().optional()
}).refine((data) => data.password === data.passwordConfirmation, {
  message: "passwords don't match",
  path: ["passwordConfirmation"]
}).refine((data) => {
  if (data.married && !data.partner) return false;
  return true;
}, {
  message: "partner name is required when married is true",
  path: ["partner"]
});
export type Signup = z.infer<typeof SignupSchema>;

File: 5. Validations and transformations for backend engineers/code/rust/example_5.rs
use serde::Deserialize;
use validator::{Validate, ValidationError};

#[derive(Debug, Deserialize, Validate)]
#[validate(schema(function = "cross_field_rules", skip_on_field_errors = false))]
pub struct Signup {
    #[validate(length(min = 8))]
    pub password: String,
    pub password_confirmation: String,
    pub married: bool,
    pub partner: Option<String>,
}

fn cross_field_rules(s: &Signup) -> Result<(), ValidationError> {
    if s.password != s.password_confirmation {
        return Err(ValidationError::new("passwords don't match"));
    }
    if s.married && s.partner.is_none() {
        return Err(ValidationError::new("partner name is required when married is true"));
    }
    Ok(())
}
